File: packages/amplify_datastore/ios/Classes/types/model/FlutterModelSchema.swift
/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import Flutter
import Foundation
import Amplify

struct FlutterModelSchema {
    let name: String
    let fields: [String: FlutterModelField]
    let pluralName: String?
    let authRules: [FlutterAuthRule]?

    // Not used for now
    let attributes: [ModelAttribute] = []

    init(serializedData: [String: Any]) throws {
        guard let name = serializedData["name"] as? String else {
            throw ModelSchemaError.parse(
                className: "FlutterModelSchema",
                fieldName: "name",
                desiredType: "String")
        }
        self.name = name

        guard let inputFieldsMap = serializedData["fields"] as? [String: [String: Any]] else {
            throw ModelSchemaError.parse(
                className: "FlutterModelSchema",
                fieldName: "fields",
                desiredType: "[String: [String: Any]]")
        }
        self.fields = try inputFieldsMap.mapValues {
            try FlutterModelField.init(serializedData: $0)
        }

        self.pluralName = serializedData["pluralName"] as? String

        if let inputAuthRulesMap = serializedData["authRules"] as? [[String: Any]] {
            self.authRules = try inputAuthRulesMap.map {
                try FlutterAuthRule(serializedData: $0)
            }
        } else {
            self.authRules = nil
        }
    }

    public func convertToNativeModelSchema(customTypeSchemasRegistry: FlutterSchemaRegistry) throws -> ModelSchema {
        return ModelSchema.init(
            name: name,
            pluralName: pluralName,
            authRules: authRules?.map{
                            $0.convertToNativeAuthRule()
                        } ?? [AuthRule](),
            attributes: attributes,
            fields: try fields.mapValues { try $0.convertToNativeModelField(customTypeSchemasRegistry: customTypeSchemasRegistry) }
        )
    }
}

File: packages/datastore/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterModelSchema.kt
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.amazonaws.amplify.amplify_datastore.types.model

import com.amplifyframework.core.model.ModelSchema
import com.amplifyframework.core.model.SerializedModel

data class FlutterModelSchema(val map: Map<String, Any>) {
    val name: String = map["name"] as String
    private val pluralName: String? = map["pluralName"] as String?
    private val authRules: List<FlutterAuthRule>? =
        (map["authRules"] as List<Map<String, Any>>?)?.map { FlutterAuthRule(it) }
    private val fields: Map<String, FlutterModelField> =
        (map["fields"] as Map<String, Any>).mapValues { FlutterModelField(it.value as Map<String, Any>) }
    private val associations: Map<String, FlutterModelAssociation> =
        fields.filterKeys { key -> fields[key]?.getModelAssociation() != null }
            .mapValues { it.value.getModelAssociation()!! }

    fun convertToNativeModelSchema(): ModelSchema {
        var builder: ModelSchema.Builder = ModelSchema.builder()
            .name(name)
            .pluralName(pluralName)
            .fields(fields.mapValues { it.value.convertToNativeModelField() })

        if (!authRules.isNullOrEmpty()) {
            builder = builder.authRules(authRules.map { it.convertToNativeAuthRule() })
        }
        if (!associations.isNullOrEmpty()) {
            builder = builder.associations(
                associations.mapValues { it.value.convertToNativeModelAssociation() }
            )
        }
        builder.modelClass(SerializedModel::class.java)
        return builder.build()
    }
}
